### PR TITLE
Accessibility guidelines

### PR DESCRIPTION
**Pull request by robertknight**
This is an initial stab at a set of guidelines for making web content / interfaces in our products accessible to a wider range of users.

It is intended to be useful as a shared reference to consult when implementing new features, or reviewing changes. The intended audience includes Hypothesis devs (current and new ones, those who work mostly on the frontend and those who only occasionally make front-end changes) and external contributors. I am assuming basic web development familiarity.

It is not intended to be exhaustive. I've focused mostly on things that have come up regularly during code reviews and Slack discussions, as well as covering the types of change required in recent work to improve accessibility for activity pages (see https://github.com/hypothesis/h/pull/4554, https://github.com/hypothesis/h/pull/4559 and https://github.com/hypothesis/h/pull/4560).
 
Feedback on structure & content is welcome.

### CODE REVIEW

**Comment by chdorner** [comment]
I've only roughly checked it and mostly looks good to me. Based on recent Slack discussions I've requested a proper review from @fatbusinessman :)

**Comment by dhwthompson** [comment]
First impressions: yup, these all look like good points to me. As you say, it’s not exhaustive, but it’s an excellent start.

I especially like the repeated advice to use semantic markup. 👍

File: docs/accessibility.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+practices and provide a shared reference which is useful when implementing or
+reviewing UI changes.
+
+## Guidelines
+
+### Mobile
+
+1. **Use media queries to adapt layout to display size** - Use `@media` queries
+   to adjust the layout depending on screen size to provide more usable
+   interfaces on small screens.
```
Comment (seanh) [inline docs/accessibility.md:18]: One thing that could be added here - what is our approach to using media queries? For example I've seen two distinct approaches, one is to have media queries for different devices or form factors. "On iPad it should do X", "at tablet size or below it should do X", etc. Another approach I've seen is to let the boundaries of the media queries be defined by the content: "When the display size is too small to show layout X, then switch to layout Y" (media query rules involving things like pixel sizes, derived by trial and error to see at what size the layout needs to change).

File: docs/accessibility.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+1. **Make interactive controls big enough to tap.** Make sure any UI controls
+   that respond to taps are large enough to tap easily with a finger.
+ * Interactive controls should generally be at least 44x44px in size on
+   touch-screen devices.
+ * Use `@media (pointer: coarse)` to adjust the size of tap-able items
+   on devices with touch screens.
```
Comment (seanh) [inline docs/accessibility.md]: These two bullet points aren't indented far enough, breaks out of the enumerated list

**Review by seanh — APPROVED** [review]
Agree with @fatbusinessman that this is a great start.

One suggestion I'd like to make: I think this accessibility stuff, and plenty of other things as well, could work well as part of a **code review checklist** that can be referred to by both the author and the reviewer of a pull request.

File: docs/accessibility.md (hunk 1 above)
Comment (robertknight) [reply docs/accessibility.md:18]: The way I look at it is that content such as Activity Pages has a few major layouts which are appropriate for the typical range of screen sizes of a phone, a tablet and desktop. Those major layouts are likely to be specified by the designer. In the implementation, it is useful to have common variables which define the boundaries of these ranges that are used in the major media queries.

Given that screens at the small end of a range may have a width / height several hundred px different from screens at the large end of a range, it makes sense to adapt/refine the main layout or the layout of individual elements where possible to make the best use of available space. Much of this adaptation can be done just relying on flexbox without any media queries, but there are cases where additional breakpoints are needed. These smaller refinements may be things that we (developers) can do ourselves without needing upfront specification from the person working on the design. An example would be the way that some of the margins disappear in activity pages as the screen size gets smaller towards the lower end of one of the ranges.

Does this answer your question?

**Comment by robertknight** [comment]
> I think this accessibility stuff, and plenty of other things as well, could work well as part of a code review checklist that can be referred to by both the author and the reviewer of a pull request.

👍 good suggestion. I had considered some kind of "UI component checklist" as a shared reference when building some part of the frontend. By "component" here I mean some logical piece of the page / app, not referring to any specific framework.

**Pull request merged**